src/main/java/co/cask/hydrator/salesforce/plugin/source/batch/SalesforceBatchSource.java

```diff
@@ -0,0 +1,264 @@
+
+    if (schema != null) {
+      if (schema.getFields() != null) {
+        lineageRecorder.recordRead("Read", "Read from Salesforce",
+                                   schema.getFields().stream().map(Schema.Field::getName)
+                                     .collect(Collectors.toList()));
+      }
+    }
+  }
+
+  @Override
+  public void transform(KeyValue<String, String> input,
+                        Emitter<StructuredRecord> emitter) throws Exception {
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+
+    String[] fieldNames = getValuesFromCSVRow(input.getKey());
+    String[] values = getValuesFromCSVRow(input.getValue());
+
+    if (fieldNames.length != values.length) {
+      throw new IllegalArgumentException("Number of fields is not equal to the number of values");
+    }
+
+    for (int i = 0; i < fieldNames.length; i++) {
+      String fieldName = fieldNames[i];
+      String value = values[i];
+
+      Schema.Field field = schema.getField(fieldName);
+
+      if (field == null) {
+        continue; // this field is not in schema
+      }
+
+      builder.set(fieldName, convertValue(value, field));
+    }
+
+    emitter.emit(builder.build());
+  }
+
+  /**
+   * Request object for retrieving schema from the Salesforce
+   */
+  class Request {
+    String query;
+    String username;
+    String password;
+    String clientId;
+    String clientSecret;
+    String loginUrl;
+  }
+
+  /**
+   * Get Salesforce schema by query.
+   *
+   * @param request request with credentials and query
+   * @return schema calculated from query
+   * @throws Exception is thrown by httpclientlib
+   */
+  @Path("getSchema")
+  public Schema getSchema(Request request) throws Exception {
+    return SalesforceSchemaUtil.getSchemaFromQuery(new AuthenticatorCredentials(request.username, request.password,
+                                                                                request.clientId, request.clientSecret,
+                                                                                request.loginUrl), request.query);
+  }
+
+  /**
+   * An advanced version of split csv row by comma. We cannot simply split by comma since some values may have comma
+   * inside them and Salesforce does not escape it.
+   *
+   * The format of Salesforce csv row is:
+   * "value1","value2","value3"
+   *
+   * So we are splitting by "\",\"" instead of by comma. Since quotes are not allowed without escaping
+   * in values this works in any case.
+   *
+   *
+   * @param csvRow one row in csv format with quoted values
+   * @return an array of values
+   */
+  private String[] getValuesFromCSVRow(String csvRow) {
+    String[] values = csvRow.split("\",\"");
+
+    values[0] = values[0].substring(1);
+
+    String last = values[values.length - 1];
+    values[values.length - 1] = last.substring(0, last.length() - 1);
+
+    return values;
+  }
+
+  private Object convertValue(String value, Schema.Field field) {
```

Comment:
is it possible for value to be null here?

Reply:
I don't think so. Value comes parsed out of csv split. So only strings will be here.